freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/user/UserSyncService.java

```diff
@@ -413,181 +455,151 @@ UsersState getIpaUserState(FreeIpaClient freeIpaClient, UmsUsersState umsUsersSt
             try {
-                com.sequenceiq.freeipa.client.model.User userAdd = freeIpaClient.userAdd(
-                        username, fmsUser.getFirstName(), fmsUser.getLastName());
-                LOGGER.debug("Success: {}", userAdd);
+                handleOperation(operation, operations, fmsToFreeipaBatchCallEnabled, freeIpaClient);
             } catch (FreeIpaClientException e) {
                 if (FreeIpaClientExceptionUtil.isDuplicateEntryException(e)) {
                     LOGGER.debug("user '{}' already exists", fmsUser.getName());
                 } else {
-                    LOGGER.error("Failed to add {}", username, e);
+                    LOGGER.error("Failed to add {}", fmsUser.getName(), e);
                     warnings.accept(fmsUser.getName(), "Failed to add user:" + e.getMessage());
-                    checkIfClientStillUsable(e);
+                    freeIpaClient.checkIfClientStillUsable(e);
                 }
             }
         }
+        if (fmsToFreeipaBatchCallEnabled) {
+            freeIpaClient.callBatch(warnings, operations, Set.of(FreeIpaErrorCodes.DUPLICATE_ENTRY));
+        }
     }
 
-    private void removeUsers(FreeIpaClient freeIpaClient, Set<String> fmsUsers, BiConsumer<String, String> warnings) throws FreeIpaClientException {
-        for (String username : fmsUsers) {
-            LOGGER.debug("Removing user {}", username);
+    void removeUsers(boolean fmsToFreeipaBatchCallEnabled, FreeIpaClient freeIpaClient, Set<String> fmsUsers,
+            BiConsumer<String, String> warnings) throws FreeIpaClientException {
+        List<Object> operations = Lists.newArrayList();
+        for (String user : fmsUsers) {
+            UserRemoveOperation operation = UserRemoveOperation.get(user);
             try {
-                com.sequenceiq.freeipa.client.model.User userRemove = freeIpaClient.deleteUser(username);
-                LOGGER.debug("Success: {}", userRemove);
+                handleOperation(operation, operations, fmsToFreeipaBatchCallEnabled, freeIpaClient);
             } catch (FreeIpaClientException e) {
                 if (FreeIpaClientExceptionUtil.isNotFoundException(e)) {
-                    LOGGER.debug("user '{}' already does not exists", username);
+                    LOGGER.debug("user '{}' does not exists", user);
                 } else {
-                    LOGGER.error("Failed to delete {}", username, e);
-                    warnings.accept(username, "Failed to remove user:" + e.getMessage());
-                    checkIfClientStillUsable(e);
+                    LOGGER.error("Failed to delete {}", user, e);
+                    warnings.accept(user, "Failed to remove user:" + e.getMessage());
+                    freeIpaClient.checkIfClientStillUsable(e);
                 }
             }
         }
+        if (fmsToFreeipaBatchCallEnabled) {
+            freeIpaClient.callBatch(warnings, operations, Set.of(FreeIpaErrorCodes.NOT_FOUND));
+        }
```

Comment:
I see these 3 lines multiple times and also the `handleOperation` has the same if condition. It would be possible to collect the operations at first, and if it's batch then call the batch API else invoke them 1-by-1. In this case handleOperation won't invoke it, just collect them, and there could be an invoke/call part with the if condition.

Reply:
unfortunately, there is two types of callBatch, one with default partition size and one for custom partition size, also some operation have one foreach loop, some has two foreach loop, so each operation differ a little bit. 
so in case of your suggestion, collection part of the logic should be somehow generalized which is quite hard in case of these operations, since for loops are quite custom, but if there is a way, let's discuss it in private.

Reply:
ok, for second thought, i see your point, yes, i think it will work